enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/PresenceCheckService.java

```diff
@@ -341,9 +366,27 @@ private IdentityVerificationEntity fetchIdVerification(OwnerId ownerId) throws P
     }
 
     private SessionInfo updateSessionInfo(final OwnerId ownerId, final IdentityVerificationEntity identityVerification, final SessionInfo sessionInfo) {
-        sessionInfo.getSessionAttributes().put("timestampLastUsed", ownerId.getTimestamp());
+        sessionInfo.getSessionAttributes().put(SESSION_ATTRIBUTE_TIMESTAMP_LAST_USED, ownerId.getTimestamp());
         identityVerification.setSessionInfo(jsonSerializationService.serialize(sessionInfo));
         return sessionInfo;
     }
 
+    /**
+     * Return whether the image was already uploaded.
+     * <p>
+     * The logic is based on the fact, that the given verification identity may have several attempts for presence check
+     * and when passed {@link #checkPresenceVerification}, specific attributes of {@link SessionInfo} were set.
+     *
+     * @param identityVerification Verification identity.
+     * @return true if image was already uploaded
+     */
+    private boolean wasImageAlreadyUploaded(final IdentityVerificationEntity identityVerification) {
+        final String sessionInfoString = identityVerification.getSessionInfo();
+        if (!StringUtils.hasText(sessionInfoString)) {
+            return false;
+        }
+        final SessionInfo sessionInfo = jsonSerializationService.deserialize(sessionInfoString, SessionInfo.class);
+        // TODO (racansky, 2023-05-31) improve logic to something more robust than presence of the attribute `timestampLastUsed`
+        return sessionInfo != null && !CollectionUtils.isEmpty(sessionInfo.getSessionAttributes()) && sessionInfo.getSessionAttributes().containsKey(SESSION_ATTRIBUTE_TIMESTAMP_LAST_USED);
```

Comment:
Is there any reason why we just don't use some new property e.g. `imageUploaded`:`true` for this?

Reply:
😞 Indeed, I have a TODO there. Let's improve it later.

Reply:
We need to improve it now. It is not clear how the session use attribute is propagated and if this is relevant for the image upload at all.

Reply:
Ok, I have refactored the way how we work with the `SessionInfo`.